File: Java/ObserverPattern/src/com/laxmena/NewsAgency.java
package com.laxmena;

import java.util.ArrayList;
import java.util.List;

/**
 * There are two types of implementation in Publisher Subscriber models - Push and Pull models.
 *
 * In Push model, whenever there is an update, the publisher pushes the update to the subscriber.
 * In Pull model, publisher notifies the subscriber that there is an update, and subscriber decides
 * when to pull the data.
 *
 * In this example, we implement the Pull model.
 */

public abstract class NewsAgency implements NewsPublisher {
    String publisher;
    String name;
    ArrayList<NewsSubscriber> subscribers;
    ArrayList<News> newsList;

    public NewsAgency() {
        subscribers = new ArrayList<NewsSubscriber>();
        newsList = new ArrayList<News>();
    }

    @Override
    public void registerSubscriber(NewsSubscriber subscriber) {
        this.subscribers.add(subscriber);
    }

    @Override
    public void removeSubscriber(NewsSubscriber subscriber) {
        this.subscribers.remove(subscriber);
    }

    @Override
    public void notifySubscribers() {
        for(NewsSubscriber subscriber: this.subscribers) {
            // Subscriber can subscribe to multiple publishers.
            // So we send a reference to the subscriber.
            subscriber.update(publisher);
        }
    }

    @Override
    public News getNews(int newsId) {
        if(newsId < 0 || newsId >= this.newsList.size())
            return null;
        return this.newsList.get(newsId);
    }
}

File: Observer/Python/NewsAgency/NewsAgency/news_agency.py
import string
from abc import ABCMeta
from datetime import datetime
from typing import List

from News.news import News
from interface.news_publisher import NewsPublisher
from interface.news_subscriber import NewsSubscriber


class NewsAgency(NewsPublisher, metaclass=ABCMeta):
    publisher: string = None
    name: string = None
    subscribers: List[NewsSubscriber] = []
    news_list: List[News] = []

    def register_subscriber(self, subscriber: NewsSubscriber):
        if subscriber not in self.subscribers:
            self.subscribers.append(subscriber)

    def unsubscribe(self, subscriber: NewsSubscriber):
        if subscriber in self.subscribers:
            self.subscribers.remove(subscriber)

    def notify_subscribers(self):
        news_id = len(self.news_list) - 1
        for subscriber in self.subscribers:
            subscriber.update(news_id)

    def add_news(self, title, content):
        news = News(title, content, datetime.now().strftime("%m/%d/%Y, %H:%M:%S"), self.publisher)
        self.news_list.append(news)
        self.notify_subscribers()

    def get_news(self, news_id):
        if news_id < 0 or news_id >= len(self.news_list):
            return None
        return self.news_list[news_id]
